File: source/modulo_components/test/cpp/test_component_interface_parameters.cpp
#include <gtest/gtest.h>

#include "modulo_components/exceptions/ComponentParameterException.hpp"
#include "modulo_core/EncodedState.hpp"

#include "test_modulo_components/component_public_interfaces.hpp"

namespace modulo_components {

template<class NodeT>
class ComponentInterfaceParameterTest : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    rclcpp::init(0, nullptr);
  }

  static void TearDownTestSuite() {
    rclcpp::shutdown();
  }

  void SetUp() override {
    if (std::is_same<NodeT, rclcpp::Node>::value) {
      this->component_ = std::make_shared<ComponentInterfacePublicInterface<NodeT>>(
          rclcpp::NodeOptions(), modulo_core::communication::PublisherType::PUBLISHER
      );
    } else if (std::is_same<NodeT, rclcpp_lifecycle::LifecycleNode>::value) {
      this->component_ = std::make_shared<ComponentInterfacePublicInterface<NodeT>>(
          rclcpp::NodeOptions(), modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER
      );
    }
    param_ = state_representation::make_shared_parameter("test", 1);
  }

  template<typename T>
  void expect_parameter_value(const T& value) {
    EXPECT_STREQ(this->component_->get_ros_parameter("test").value_to_string().c_str(), std::to_string(value).c_str());
    EXPECT_EQ(this->component_->template get_parameter_value<T>("test"), value);
    EXPECT_EQ(this->component_->parameter_map_.template get_parameter_value<T>("test"), value);
  }

  std::shared_ptr<ComponentInterfacePublicInterface<NodeT>> component_;
  std::shared_ptr<state_representation::Parameter<int>> param_;
};
using NodeTypes = ::testing::Types<rclcpp::Node, rclcpp_lifecycle::LifecycleNode>;
TYPED_TEST_SUITE(ComponentInterfaceParameterTest, NodeTypes);

TYPED_TEST(ComponentInterfaceParameterTest, AddParameter) {
  EXPECT_THROW(auto discard = this->component_->get_parameter("test"), exceptions::ComponentParameterException);
  EXPECT_THROW(this->component_->get_ros_parameter("test"), rclcpp::exceptions::ParameterNotDeclaredException);
  this->component_->add_parameter(this->param_, "Test parameter");

  // Adding the parameter should declare and set the value and call the validation function
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  EXPECT_NO_THROW(auto discard = this->component_->get_parameter("test"));
  EXPECT_NO_THROW(this->component_->get_ros_parameter("test"));
  this->template expect_parameter_value<int>(1);
}

TYPED_TEST(ComponentInterfaceParameterTest, AddNameValueParameter) {
  EXPECT_THROW(auto discard = this->component_->get_parameter("test"), exceptions::ComponentParameterException);
  EXPECT_THROW(this->component_->get_ros_parameter("test"), rclcpp::exceptions::ParameterNotDeclaredException);
  this->component_->add_parameter("test", 1, "Test parameter");

  // Adding the parameter should declare and set the value and call the validation function
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  EXPECT_NO_THROW(auto discard = this->component_->get_parameter("test"));
  EXPECT_NO_THROW(this->component_->get_ros_parameter("test"));
  this->template expect_parameter_value<int>(1);
}

TYPED_TEST(ComponentInterfaceParameterTest, AddParameterAgain) {
  EXPECT_THROW(auto discard = this->component_->get_parameter("test"), exceptions::ComponentParameterException);
  EXPECT_THROW(this->component_->get_ros_parameter("test"), rclcpp::exceptions::ParameterNotDeclaredException);
  this->component_->add_parameter(this->param_, "Test parameter");

  // Adding an existing parameter again should just set the value
  this->component_->validate_parameter_was_called = false;
  EXPECT_NO_THROW(this->component_->add_parameter("test", 2, "foo"));
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  this->template expect_parameter_value<int>(2);
  EXPECT_EQ(this->param_->get_value(), 2);
}

TYPED_TEST(ComponentInterfaceParameterTest, SetParameter) {
  // setting before adding should not work and parameter should not be created
  this->component_->set_parameter_value("test", 1);
  EXPECT_THROW(this->component_->template get_parameter_value<int>("test"), exceptions::ComponentParameterException);

  // validation should not be invoked as the parameter did not exist
  EXPECT_FALSE(this->component_->validate_parameter_was_called);
  EXPECT_THROW(this->component_->get_ros_parameter("test"), rclcpp::exceptions::ParameterNotDeclaredException);

  this->component_->add_parameter(this->param_, "Test parameter");

  // Setting the parameter value should call the validation function and update the referenced value
  this->component_->validate_parameter_was_called = false;
  EXPECT_NO_THROW(this->component_->set_parameter_value("test", 2));
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  this->template expect_parameter_value<int>(2);
  EXPECT_EQ(this->param_->get_value(), 2);

  // If the validation function returns false, setting the parameter value should _not_ update the referenced value
  this->component_->validate_parameter_was_called = false;
  this->component_->validate_parameter_return_value = false;
  this->component_->set_parameter_value("test", 3);
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  this->template expect_parameter_value<int>(2);
  EXPECT_EQ(this->param_->get_value(), 2);

  // Setting a value with an incompatible type should not update the parameter
  this->component_->template set_parameter_value<std::string>("test", "foo");
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  this->template expect_parameter_value<int>(2);
  EXPECT_EQ(this->param_->get_value(), 2);
}

TYPED_TEST(ComponentInterfaceParameterTest, SetParameterROS) {
  this->component_->add_parameter(this->param_, "Test parameter");

  // Setting the parameter value should call the validation function and update the referenced value
  this->component_->validate_parameter_was_called = false;
  auto result = this->component_->set_ros_parameter({"test", 2});
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  EXPECT_TRUE(result.successful);
  this->template expect_parameter_value<int>(2);
  EXPECT_EQ(this->param_->get_value(), 2);

  // If the validation function returns false, setting the parameter value should _not_ update the referenced value
  this->component_->validate_parameter_was_called = false;
  this->component_->validate_parameter_return_value = false;
  result = this->component_->set_ros_parameter({"test", 3});
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  EXPECT_FALSE(result.successful);
  this->template expect_parameter_value<int>(2);
  EXPECT_EQ(this->param_->get_value(), 2);
}

TYPED_TEST(ComponentInterfaceParameterTest, GetParameterDescription) {
  this->component_->add_parameter(this->param_, "Test parameter");
  EXPECT_STREQ(this->component_->get_parameter_description("test").c_str(), "Test parameter");

  EXPECT_THROW(this->component_->get_parameter_description("foo"), rclcpp::exceptions::ParameterNotDeclaredException);
}

TYPED_TEST(ComponentInterfaceParameterTest, ReadOnlyParameter) {
  // Adding a read-only parameter should behave normally
  this->component_->add_parameter(this->param_, "Test parameter", true);
  EXPECT_TRUE(this->component_->validate_parameter_was_called);
  EXPECT_NO_THROW(auto discard = this->component_->get_parameter("test"));
  EXPECT_NO_THROW(this->component_->get_ros_parameter("test"));
  this->template expect_parameter_value<int>(1);

  // Trying to set the value of the read-only parameter should fail before the validation step
  this->component_->validate_parameter_was_called = false;
  this->component_->set_parameter_value("test", 2);
  EXPECT_FALSE(this->component_->validate_parameter_was_called);
  this->template expect_parameter_value<int>(1);
  EXPECT_EQ(this->param_->get_value(), 1);

  // Setting the value on the ROS interface should also fail
  this->component_->validate_parameter_was_called = false;
  auto result = this->component_->set_ros_parameter({"test", 2});
  EXPECT_FALSE(this->component_->validate_parameter_was_called);
  EXPECT_FALSE(result.successful);
  this->template expect_parameter_value<int>(1);
  EXPECT_EQ(this->param_->get_value(), 1);
}
} // namespace modulo_components

File: source/modulo_components/test/python/test_component_interface_parameters.py
import pytest
import rclpy.exceptions
import rclpy.node
import state_representation as sr
from modulo_components.component_interface import ComponentInterface
from rcl_interfaces.msg import SetParametersResult

from modulo_components.exceptions.component_exceptions import ComponentParameterError


class ComponentInterfaceTest(ComponentInterface):
    def __init__(self, node_name, *kargs, **kwargs):
        self.validation_return_value = True
        super().__init__(node_name, *kargs, **kwargs)
        self.validate_was_called = False
        self.param = sr.Parameter("test", 1, sr.ParameterType.INT)

    def get_ros_parameter(self, name: str) -> rclpy.Parameter:
        return rclpy.node.Node.get_parameter(self, name)

    def set_ros_parameter(self, param: rclpy.Parameter) -> SetParametersResult:
        return rclpy.node.Node.set_parameters(self, [param])[0]

    def _validate_parameter(self, parameter: sr.Parameter) -> bool:
        self.validate_was_called = True
        return self.validation_return_value


@pytest.fixture()
def component_interface(ros_context):
    yield ComponentInterfaceTest('component_interface')


def assert_param_value_equal(component_interface: ComponentInterfaceTest, name: str, value: float):
    assert component_interface.get_ros_parameter(name).get_parameter_value().integer_value == value
    assert component_interface.get_parameter_value(name) == value
    assert component_interface.get_parameter(name).get_value() == value


def test_add_parameter(component_interface):
    with pytest.raises(ComponentParameterError):
        component_interface.get_parameter("test")
    with pytest.raises(rclpy.exceptions.ParameterNotDeclaredException):
        rclpy.node.Node.get_parameter(component_interface, "test")

    component_interface.add_parameter("param", "Test parameter")
    assert component_interface.validate_was_called
    component_interface.get_parameter("test")
    rclpy.node.Node.get_parameter(component_interface, "test")
    assert component_interface.param.get_parameter_type() == sr.ParameterType.INT
    assert component_interface.param.get_value() == 1


def test_add_parameter_again(component_interface):
    with pytest.raises(ComponentParameterError):
        component_interface.get_parameter_value("test")
    with pytest.raises(rclpy.exceptions.ParameterNotDeclaredException):
        rclpy.node.Node.get_parameter(component_interface, "test")

    component_interface.add_parameter("param", "Test parameter")
    component_interface.validate_was_called = False
    component_interface.add_parameter(sr.Parameter("test", 2, sr.ParameterType.INT), "foo")
    assert component_interface.validate_was_called
    assert_param_value_equal(component_interface, "test", 2)
    assert component_interface.param.get_value() == 2


def test_add_parameter_again_not_attribute(component_interface):
    with pytest.raises(ComponentParameterError):
        component_interface.get_parameter("test")
    with pytest.raises(rclpy.exceptions.ParameterNotDeclaredException):
        rclpy.node.Node.get_parameter(component_interface, "test")

    component_interface.add_parameter(component_interface.param, "Test parameter")
    component_interface.validate_was_called = False
    component_interface.add_parameter(sr.Parameter("test", 2, sr.ParameterType.INT), "foo")
    assert component_interface.validate_was_called
    assert_param_value_equal(component_interface, "test", 2)
    assert component_interface.param.get_value() == 1


def test_set_parameter(component_interface):
    assert not component_interface.validate_was_called
    with pytest.raises(rclpy.exceptions.ParameterNotDeclaredException):
        component_interface.set_parameter_value("test", 1, sr.ParameterType.INT)
    assert not component_interface.validate_was_called
    with pytest.raises(rclpy.exceptions.ParameterNotDeclaredException):
        component_interface.get_ros_parameter("test")

    component_interface.add_parameter("param", "Test parameter")

    component_interface.validate_was_called = False
    component_interface.set_parameter_value("test", 2, sr.ParameterType.INT)
    assert component_interface.validate_was_called
    assert_param_value_equal(component_interface, "test", 2)
    assert component_interface.param.get_value() == 2

    component_interface.validate_was_called = False
    component_interface.validation_return_value = False
    with pytest.raises(RuntimeError):
        component_interface.set_parameter_value("test", 3, sr.ParameterType.INT)
    assert component_interface.validate_was_called
    assert_param_value_equal(component_interface, "test", 2)
    assert component_interface.param.get_value() == 2

    with pytest.raises(RuntimeError):
        component_interface.set_parameter_value("test", "test", sr.ParameterType.STRING)
    assert component_interface.validate_was_called
    assert_param_value_equal(component_interface, "test", 2)
    assert component_interface.param.get_value() == 2


def test_set_parameter_ros(component_interface):
    component_interface.add_parameter("param", "Test parameter")

    component_interface.validate_was_called = False
    result = component_interface.set_ros_parameter(rclpy.Parameter("test", value=2))
    assert component_interface.validate_was_called
    assert result.successful
    assert_param_value_equal(component_interface, "test", 2)
    assert component_interface.param.get_value() == 2

    component_interface.validate_was_called = False
    component_interface.validation_return_value = False
    result = component_interface.set_ros_parameter(rclpy.Parameter("test", value=3))
    assert component_interface.validate_was_called
    assert not result.successful
    assert_param_value_equal(component_interface, "test", 2)
    assert component_interface.param.get_value() == 2


def test_get_parameter_description(component_interface):
    component_interface.add_parameter(component_interface.param, "Test parameter")
    assert component_interface.describe_parameter("test").description == "Test parameter"

    with pytest.raises(rclpy.exceptions.ParameterNotDeclaredException):
        component_interface.describe_parameter("foo")


def test_read_only_parameter(component_interface):
    component_interface.add_parameter("param", "Test parameter", True)
    assert component_interface.validate_was_called
    component_interface.get_parameter("test")
    component_interface.get_ros_parameter("test")
    assert_param_value_equal(component_interface, "test", 1)

    # TODO read only
    component_interface.validate_was_called = False
    # with pytest.raises(RuntimeError):
    #     component_interface.set_parameter_value("test", 2, sr.ParameterType.INT)
    # assert not component_interface.validate_was_called
    assert_param_value_equal(component_interface, "test", 1)
    assert component_interface.param.get_value() == 1

    component_interface.validate_was_called = False
    # result = component_interface.set_ros_parameter(rclpy.Parameter("test", value=2))
    # assert not component_interface.validate_was_called
    # assert not result.successful
    assert_param_value_equal(component_interface, "test", 1)
